File: backend/src/utils/xlsxHelper.ts
import ExcelJS from 'exceljs';
import { Response } from 'express';

// ── palette ─────────────────────────────────────────────────────────────────
const PRIMARY   = 'FF4338CA'; // indigo-700
const HDR_BG    = 'FFE0E7FF'; // indigo-100
const ALT_ROW   = 'FFFAFAFA';
const WHITE     = 'FFFFFFFF';

type Fill = ExcelJS.Fill;
const fill = (argb: string): Fill => ({ type: 'pattern', pattern: 'solid', fgColor: { argb } });

// ── public helpers ───────────────────────────────────────────────────────────

export const fmtDate = (d?: string | Date | null): string => {
  if (!d) return '—';
  return new Date(d).toLocaleDateString('es-AR', { day: '2-digit', month: '2-digit', year: '2-digit' });
};

export const estadoLabel: Record<string, string> = {
  activo: 'Activo', baja: 'Baja', en_reparacion: 'En reparación', almacenado: 'En depósito',
  abierto: 'Abierto', en_progreso: 'En progreso', cerrado: 'Cerrado',
};
export const estadoColor: Record<string, string> = {
  activo: 'FFD1FAE5', baja: 'FFFEE2E2', en_reparacion: 'FFFED7AA', almacenado: 'FFEDE9FE',
  abierto: 'FFDBEAFE', en_progreso: 'FFFED7AA', cerrado: 'FFF3F4F6',
};
export const prioColor: Record<string, string> = {
  alta: 'FFFEE2E2', media: 'FFFEF9C3', baja: 'FFD1FAE5',
};
export const factColor: Record<string, string> = {
  pendiente_cobro: 'FFFEF9C3', facturada: 'FFD1FAE5', sin_facturar: 'FFF3F4F6',
};
export const factLabel: Record<string, string> = {
  pendiente_cobro: 'Pend. cobro', facturada: 'Facturada',
};
export const movLabel: Record<string, string> = {
  alta: 'Alta', baja: 'Baja', traslado: 'Traslado',
  en_reparacion: 'Reparación', almacenado: 'Almacenado', activo: 'Reactivado',
};

// ── workbook setup ───────────────────────────────────────────────────────────

export interface XlsxMeta { titulo: string; filtros: string[]; generadoEl: string }

export function createSheet(meta: XlsxMeta, sheetName: string) {
  const wb = new ExcelJS.Workbook();
  wb.creator = 'IT Support Hub';
  wb.created = new Date();
  const ws = wb.addWorksheet(sheetName, {
    pageSetup: { paperSize: 9, orientation: 'landscape', fitToPage: true, fitToWidth: 1, fitToHeight: 0 },
  });
  return { wb, ws };
}

export function addMetaHeader(ws: ExcelJS.Worksheet, meta: XlsxMeta, colCount: number): number {
  ws.mergeCells(1, 1, 1, colCount);
  const r1 = ws.getCell('A1');
  r1.value = `IT Support Hub — ${meta.titulo}`;
  r1.font = { bold: true, size: 13, color: { argb: WHITE } };
  r1.fill = fill(PRIMARY);
  r1.alignment = { vertical: 'middle', horizontal: 'left', indent: 1 };
  ws.getRow(1).height = 26;

  const infoText = [`Generado: ${meta.generadoEl}`, ...meta.filtros].join('   |   ');
  ws.mergeCells(2, 1, 2, colCount);
  const r2 = ws.getCell('A2');
  r2.value = infoText;
  r2.font = { size: 9, italic: true, color: { argb: 'FF6B7280' } };
  r2.fill = fill('FFF1F5F9');
  r2.alignment = { vertical: 'middle', horizontal: 'left', indent: 1 };
  ws.getRow(2).height = 15;

  ws.getRow(3).height = 5;
  return 4;
}

export function addHeaderRow(ws: ExcelJS.Worksheet, rowNum: number, headers: string[], widths: number[]) {
  const row = ws.getRow(rowNum);
  headers.forEach((h, i) => {
    const cell = row.getCell(i + 1);
    cell.value = h;
    cell.font = { bold: true, size: 10, color: { argb: 'FF1E1B4B' } };
    cell.fill = fill(HDR_BG);
    cell.border = { bottom: { style: 'thin', color: { argb: 'FFC7D2FE' } } };
    cell.alignment = { vertical: 'middle', horizontal: 'left' };
    ws.getColumn(i + 1).width = widths[i] || 14;
  });
  row.height = 20;
  ws.views = [{ state: 'frozen', ySplit: rowNum, activeCell: `A${rowNum + 1}` }];
}

export function addDataRow(
  ws: ExcelJS.Worksheet,
  rowNum: number,
  values: (string | number | null)[],
  cellFills?: (string | null)[],
  isAlt = false,
) {
  const rowBg = isAlt ? ALT_ROW : WHITE;
  const row = ws.getRow(rowNum);
  values.forEach((v, i) => {
    const cell = row.getCell(i + 1);
    cell.value = v ?? '—';
    cell.font = { size: 10 };
    cell.fill = fill(cellFills?.[i] ?? rowBg);
    cell.alignment = { vertical: 'middle', wrapText: false };
    cell.border = { bottom: { style: 'hair', color: { argb: 'FFEDEDED' } } };
  });
  row.height = 16;
}

export function addTotalsRow(ws: ExcelJS.Worksheet, rowNum: number, values: (string | number | null)[]) {
  const row = ws.getRow(rowNum);
  values.forEach((v, i) => {
    const cell = row.getCell(i + 1);
    cell.value = v ?? '';
    cell.font = { bold: true, size: 10, color: { argb: 'FF1E1B4B' } };
    cell.fill = fill(HDR_BG);
    cell.border = { top: { style: 'thin', color: { argb: 'FFC7D2FE' } } };
    cell.alignment = { vertical: 'middle' };
  });
  row.height = 18;
}

export function addGroupHeader(ws: ExcelJS.Worksheet, rowNum: number, text: string, colCount: number) {
  ws.mergeCells(rowNum, 1, rowNum, colCount);
  const cell = ws.getCell(rowNum, 1);
  cell.value = text;
  cell.font = { bold: true, size: 10, color: { argb: 'FF3730A3' } };
  cell.fill = fill('FFEEF2FF');
  cell.border = { top: { style: 'thin', color: { argb: 'FFC7D2FE' } } };
  cell.alignment = { vertical: 'middle', horizontal: 'left', indent: 1 };
  ws.getRow(rowNum).height = 18;
}

export async function sendXlsx(wb: ExcelJS.Workbook, res: Response, filename: string) {
  res.setHeader('Content-Type', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet');
  res.setHeader('Content-Disposition', `attachment; filename="${encodeURIComponent(filename)}"`);
  await wb.xlsx.write(res);
  res.end();
}

File: app/services/xlsx_helper.py
"""Helpers de export a Excel con openpyxl — reconstruye el diseno visual
de `xlsxHelper.ts` (backend Node.js viejo) en Python: paleta indigo,
encabezado con titulo+filtros, header congelado, filas alternadas,
resaltado por celda, fila de totales y encabezados de grupo. No hay
precedente de export xlsx en la familia Libra (Gestiolibra no exporta) —
es una pieza propia de LibraDesk.

**Solo el como, no el que.** Las etiquetas de estado/prioridad/cobro y los
colores de resaltado vivian aca hasta el 2026-08-04; se mudaron a
`reporte_vista.py` cuando los reportes empezaron a verse tambien en pantalla,
porque dejaron de ser del Excel. Este modulo quedo con lo que si es suyo:
poner celdas en una hoja. Quien lo llama es `reporte_xlsx.py`."""
from datetime import datetime
from io import BytesIO

from fastapi.responses import StreamingResponse
from openpyxl import Workbook
from openpyxl.styles import Alignment, Border, Font, PatternFill, Side
from openpyxl.utils import get_column_letter
from openpyxl.worksheet.worksheet import Worksheet

PRIMARY = "FF4338CA"  # indigo-700
HDR_BG = "FFE0E7FF"  # indigo-100
ALT_ROW = "FFFAFAFA"
WHITE = "FFFFFFFF"
GROUP_BG = "FFEEF2FF"
GROUP_FG = "FF3730A3"
HDR_FG = "FF1E1B4B"
BORDE_HDR = "FFC7D2FE"

_HAIR = Border(bottom=Side(style="hair", color="FFEDEDED"))
_BOTTOM_HDR = Border(bottom=Side(style="thin", color=BORDE_HDR))
_TOP_HDR = Border(top=Side(style="thin", color=BORDE_HDR))


def _fill(argb: str) -> PatternFill:
    return PatternFill("solid", fgColor=argb)


def create_sheet(titulo: str, filtros: list[str], sheet_name: str | None = None):
    """Devuelve (wb, ws) con el encabezado de titulo+filtros ya puesto.
    La primera fila util (la de headers) es la 4."""
    wb = Workbook()
    wb.properties.creator = "LibraDesk"
    ws = wb.active
    ws.title = (sheet_name or titulo)[:31]
    # Apaisado y ajustado al ancho: estos reportes tienen muchas columnas.
    ws.page_setup.orientation = "landscape"
    ws.page_setup.fitToWidth = 1
    ws.page_setup.fitToHeight = 0
    ws.sheet_properties.pageSetUpPr.fitToPage = True
    return wb, ws


def add_meta_header(ws: Worksheet, titulo: str, filtros: list[str], col_count: int) -> int:
    ws.merge_cells(start_row=1, start_column=1, end_row=1, end_column=col_count)
    c1 = ws.cell(row=1, column=1, value=f"LibraDesk — {titulo}")
    c1.font = Font(bold=True, size=13, color=WHITE)
    c1.fill = _fill(PRIMARY)
    c1.alignment = Alignment(vertical="center", horizontal="left", indent=1)
    ws.row_dimensions[1].height = 26

    info = "   |   ".join(
        [f"Generado: {datetime.now().strftime('%d-%m-%Y %H:%M')}", *filtros]
    )
    ws.merge_cells(start_row=2, start_column=1, end_row=2, end_column=col_count)
    c2 = ws.cell(row=2, column=1, value=info)
    c2.font = Font(size=9, italic=True, color="FF6B7280")
    c2.fill = _fill("FFF1F5F9")
    c2.alignment = Alignment(vertical="center", horizontal="left", indent=1)
    ws.row_dimensions[2].height = 15
    ws.row_dimensions[3].height = 5
    return 4


def add_header_row(ws: Worksheet, row_num: int, headers: list[str], widths: list[int]) -> None:
    for i, h in enumerate(headers, start=1):
        cell = ws.cell(row=row_num, column=i, value=h)
        cell.font = Font(bold=True, size=10, color=HDR_FG)
        cell.fill = _fill(HDR_BG)
        cell.border = _BOTTOM_HDR
        cell.alignment = Alignment(vertical="center", horizontal="left")
        ws.column_dimensions[get_column_letter(i)].width = (
            widths[i - 1] if i - 1 < len(widths) else 14
        )
    ws.row_dimensions[row_num].height = 20
    ws.freeze_panes = f"A{row_num + 1}"


def add_data_row(
    ws: Worksheet,
    row_num: int,
    values: list,
    cell_fills: list[str | None] | None = None,
    is_alt: bool = False,
) -> None:
    row_bg = ALT_ROW if is_alt else WHITE
    for i, v in enumerate(values, start=1):
        cell = ws.cell(row=row_num, column=i, value=v if v is not None else "—")
        cell.font = Font(size=10)
        propio = cell_fills[i - 1] if cell_fills and i - 1 < len(cell_fills) else None
        cell.fill = _fill(propio or row_bg)
        cell.alignment = Alignment(vertical="center")
        cell.border = _HAIR
    ws.row_dimensions[row_num].height = 16


def add_totals_row(ws: Worksheet, row_num: int, values: list) -> None:
    for i, v in enumerate(values, start=1):
        cell = ws.cell(row=row_num, column=i, value=v if v is not None else "")
        cell.font = Font(bold=True, size=10, color=HDR_FG)
        cell.fill = _fill(HDR_BG)
        cell.border = _TOP_HDR
        cell.alignment = Alignment(vertical="center")
    ws.row_dimensions[row_num].height = 18


def add_group_header(ws: Worksheet, row_num: int, text: str, col_count: int) -> None:
    ws.merge_cells(start_row=row_num, start_column=1, end_row=row_num, end_column=col_count)
    cell = ws.cell(row=row_num, column=1, value=text)
    cell.font = Font(bold=True, size=10, color=GROUP_FG)
    cell.fill = _fill(GROUP_BG)
    cell.border = _TOP_HDR
    cell.alignment = Alignment(vertical="center", horizontal="left", indent=1)
    ws.row_dimensions[row_num].height = 18


def xlsx_response(wb: Workbook, filename: str) -> StreamingResponse:
    buffer = BytesIO()
    wb.save(buffer)
    buffer.seek(0)
    return StreamingResponse(
        buffer,
        media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )
